src/types.ts

```diff
@@ -700,4 +728,42 @@ export class StarknetContract {
         const func = <starknet.CairoFunction>this.abi[functionName];
         return adaptOutputUtil(rawResult, func.outputs, this.abi);
     }
+
+    /**
+     * Adapt `Event` to something more readable .
+     * @param eventNames  an array of the event's name that was emitted
+     * @param rawEvents array of  the events output as as unparsed space separated string
+     * @returns structured output
+     */
+    private adaptEvent(eventNames: string[], rawEvents: string[]) {
+        const events: DecodedEvent[] = [];
+        for (let i = 0; i < eventNames.length; i++) {
+            const event = <starknet.CairoEvent>this.abi[eventNames[i]];
+            if (!event) {
+                const msg = `Event name '${eventNames[i]}' doesn't exist on ${this.abiPath}.`;
+                throw new HardhatPluginError(PLUGIN_NAME, msg);
+            }
+            const adapted = adaptOutputUtil(rawEvents[i], event.data, this.abi);
+            events.push({ name: event.name, data: adapted });
+        }
+        return events;
+    }
+
+    /**
+     * Decode the events to a structured object with parameter names.
+     * @param events as received from the server.
+     * @returns structured object with parameter names.
+     */
+    async decodeEvents(events: starknet.Event[]): Promise<DecodedEvent[]> {
+        const rawEvents: string[] = [];
+        const eventNames: string[] = [];
+        for (const event of events) {
+            const result = event.data.map(BigInt).join(" ");
+            rawEvents.push(result);
+            // encoded event name guaranteed to be at index 0
+            const key = this.eventsSpecifications[event.keys[0]];
```

Comment:
Don't use a generic name as `key`. Call it `eventSpecification`.

Reply:
Also, do a check and throw a HardhatPluginError (for reference, see other places where this is done) if the key doesn't exist in this.eventsSpecifications.

Reply:
Name has been changed and check error is added https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/130/commits/d8ab7d892c576df6b5dd509ea54cd88f8526f065#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410a